pop-os/kernelstub
# Pop usmerge

Merge code from upstream.

This adds a couple of new features and fixes some bugs with 2.0.6

* Adds in a log life, with `-g` option to specify the log file path. Defaults to `/var/log/kernelstub.log`
* Fixes using the kernel stub loader in the nvram.
* Removes unnecessary ERROR output from successful runs when there is only one kernel installed.

Use-Testing:

Case 1:
1. Run `sudo kernelstub -v
  * Check that there is `INFO` output printed to console.
2. `cat /var/log/kernelstub.log`
  * The output in the log should be much longer, and should include `DEBUG` information.
3. Run `sudo kernelstub -g ./kernelstub.log`
  * There should be no output to console
4. `cat ./kernelstub.log`
  * The log should be present, and be filled with program output.

Case 2:
1. Delete ESP Contents: `sudo rm -r /boot/efi/EFI /boot/efi/loader`
2. Run `sudo kernelstub -vns`
  * Command should run, post output here.
3. Reboot, ensure that `Pop_OS YY.MM` is selected from the firmware boot menu.
  * System should reboot and get back into Pop_OS successfully.

Case 3:
1. Delete ESP contents as above.
2. Reboot.
  * System should fail to get back into Pop_OS
3. From live disk, install `kernelstub` package and run:
```
sudo mount /dev/root-filesystem-block-device /mnt
sudo mount /dev/esp-filesystem-block-device /mnt/boot/efi
sudo kernelstub -vmlfr "/mnt" -e "/mnt/boot/efi" -o "quiet loglevel=0"
sudo bootctl --path=/mnt/boot/efi install
```
4. Reboot
  * System should reboot into Pop.
